Added support for GQ GMC-SE

huyndao opened the pull request:
Hi there, thank you for building this nice project and framework.

I updated it with support for the GMC-SE.  It uses the RFC1201 so was fairly straightforward.  
I've tested it with my GMC-SE and everything seems to work ok.  
The only exception being that the Gyro gives a tuple of 3 numbers something like (15375, -5885, 769) however I think it is nonsensical and the numbers barely change when the device is moved.

I've updated the documentations as well.  Will create more PR in the futures if I see any significant changes needed.

Huy

Wikilicious commented on pygmc/devices/gmcse.py:
Hi, I think black would fail on this as it's expecting an empty line at the end of the file.

huyndao replied:
I just ran black on it and pushed the change up. It's in commit ef00d3e now

Wikilicious replied:
Looks like a linter error now... 
I should add a contribution guide to make this easier.

If it's easy for you to `pip install ruff` it can read the config in pyproject and lint for you.

huyndao replied:
gotcha, I just ruffed it, commit 6e93903.  Sorry about that, never used ruff before.

Wikilicious commented:
Hi @huyndao 
This is awesome! Thanks!
I really need to add a contribution guide.

huyndao commented:
What do we do about the ruff lint failure?  I don't see what is wrong with it.

Wikilicious commented:
Ok, now all linting and tests are passing...
It's just complaining that we added code but didn't add unit-tests for it.
The tests section isn't organized in the best way... 

We can do 2 paths:
1) We merge it as-is and I can add test cases later.
2) You can take a look at tests/test_devices/test_gmc320.py and make one for GMCSE (you can simply reuse the gmc320 data since it's also RFC1201)

huyndao commented:
> Ok, now all linting and tests are passing... It's just complaining that we added code but didn't add unit-tests for it. The tests section isn't organized in the best way...
> 
> We can do 2 paths:
> 
>     1. We merge it as-is and I can add test cases later.
> 
>     2. You can take a look at tests/test_devices/test_gmc320.py and make one for GMCSE (you can simply reuse the gmc320 data since it's also RFC1201)

Thanks, I will do option 2 within an hour.

Wikilicious commented:
You can do `pip install -r dev_requirements.txt` and it'll add `invoke` to your command line. You can then add a unit test and run `invoke test` and it'll show you a summary.

huyndao commented:
> You can do `pip install -r dev_requirements.txt` and it'll add `invoke` to your command line. You can then add a unit test and run `invoke test` and it'll show you a summary.

please bear with me, I'd need to get my device hooked up in order to have the correct data for the response mapping correct?

If so, it may take a bit until I finish work.

Wikilicious commented:
It's all good you're the first contributor and this is my first pypi project. 
You could get the data from your device... BUT I don't think that's needed here since `GMCSE` inherits `DeviceRFC1201` and doesn't override methods. Basically, feeding `GMC320` data into `GMCSE` class is a valid test. 
Give me a few minutes to take a look... it may just be a one liner

huyndao commented:
> It's all good you're the first contributor and this is my first pypi project. You could get the data from your device... BUT I don't think that's needed here since `GMCSE` inherits `DeviceRFC1201` and doesn't override methods. Basically, feeding `GMC320` data into `GMCSE` class is a valid test. Give me a few minutes to take a look... it may just be a one liner

okay, I pretty much reused GMC320 data except for the lines about the version, which I used GMCSE.

The tests passed locally, hopefully everything works.  

Please check commit 7793fa5.

Thanks!

Wikilicious commented:
> > It's all good you're the first contributor and this is my first pypi project. You could get the data from your device... BUT I don't think that's needed here since `GMCSE` inherits `DeviceRFC1201` and doesn't override methods. Basically, feeding `GMC320` data into `GMCSE` class is a valid test. Give me a few minutes to take a look... it may just be a one liner
> 
> okay, I pretty much reused GMC320 data except for the lines about the version, which I used GMCSE.
> 
> The tests passed locally, hopefully everything works.
> 
> Please check commit [7793fa5](https://github.com/Wikilicious/pygmc/commit/7793fa5d7b30207f42609686665997bf941a6fff).
> 
> Thanks!

Thank you for your time! 

If you want...
In README.md under Devices section, can you add a checkmark for GMC-SE (✔️=physically confirmed works)
Add a `Contributors` section after Devices section with your github link.
And... if all this wasn't enough... lol... in tests/test_connection_mock_serial.py line 135... add `pygmc.GMC` (if in linux or mac pip install mock_serial else... can only test in GitHub) 

If you're interested in the test coverage rabbit-hole... https://app.codecov.io/gh/Wikilicious/pygmc
It's a measure of lines of logic that a test touched.

More if you're curious... can see the code for github CI tests here: https://github.com/Wikilicious/pygmc/tree/master/.github/workflows

huyndao commented:
> > > It's all good you're the first contributor and this is my first pypi project. You could get the data from your device... BUT I don't think that's needed here since `GMCSE` inherits `DeviceRFC1201` and doesn't override methods. Basically, feeding `GMC320` data into `GMCSE` class is a valid test. Give me a few minutes to take a look... it may just be a one liner
> > 
> > 
> > okay, I pretty much reused GMC320 data except for the lines about the version, which I used GMCSE.
> > The tests passed locally, hopefully everything works.
> > Please check commit [7793fa5](https://github.com/Wikilicious/pygmc/commit/7793fa5d7b30207f42609686665997bf941a6fff).
> > Thanks!
> 
> Thank you for your time!
> 
> If you want... In README.md under Devices section, can you add a checkmark for GMC-SE (✔️=physically confirmed works) Add a `Contributors` section after Devices section with your github link. And... if all this wasn't enough... lol... in tests/test_connection_mock_serial.py line 135... add `pygmc.GMC` (if in linux or mac pip install mock_serial else... can only test in GitHub)
> 
> If you're interested in the test coverage rabbit-hole... https://app.codecov.io/gh/Wikilicious/pygmc It's a measure of lines of logic that a test touched.
> 
> More if you're curious... can see the code for github CI tests here: https://github.com/Wikilicious/pygmc/tree/master/.github/workflows

- I think I fixed the mock connection by mimicking what you had for GMC-320 (at least it worked here locally).
- I added the checkmark for GMC-SE as you suggested
- I also added the contributors section and the github links, I put you on top since you're the owner / main dev.  
- Coincidentally, I worked at Intel in my previous life lol.

huyndao commented:
let's try again shall we: mock_serial passed locally.

Wikilicious commented on pygmc/devices/gmcse.py:
```diff
@@ -0,0 +1,40 @@
+        else:
+            raise ConnectionError(f"Unable to connect port={port} baudrate={baudrate}")
+
+        # https://www.gqelectronicsllc.com/support/GMC_Selection_Guide.htm
```
I like the URL link

Wikilicious commented:
Well... everything is now perfect!
It's just a matter of a simple merge...
Or do we release a new version?

huyndao commented:
> Well... everything is now perfect! It's just a matter of a simple merge... Or do we release a new version?

You're the boss.  Whichever way gets it to pypi the quickest so I can use it with pip install lol.

Wikilicious commented:
> > Well... everything is now perfect! It's just a matter of a simple merge... Or do we release a new version?
> 
> You're the boss. Whichever way gets it to pypi the quickest so I can use it with pip install lol.

I'm a perpetual student...

Let's release
There's just one-more-thing™

The version.
pygmc follows semver
major.minor.patch - Look up the semver spec and see if you agree... 
I don't see this as a patch/fix for the 0.9.1 release... it's a new/added feature... 
So 0.10.0

There is a tool to do this... `bumpversion` 
`bumpversion minor`

huyndao commented:
> > > Well... everything is now perfect! It's just a matter of a simple merge... Or do we release a new version?
> > 
> > 
> > You're the boss. Whichever way gets it to pypi the quickest so I can use it with pip install lol.
> 
> I'm a perpetual student...
> 
> Let's release There's just one-more-thing™
> 
> The version. pygmc follows semver major.minor.patch - Look up the semver spec and see if you agree... I don't see this as a patch/fix for the 0.9.1 release... it's a new/added feature... So 0.10.0
> 
> There is a tool to do this... `bumpversion` `bumpversion minor`

Thanks, I'm learning as well...  Never worked with tests / CI before.

huyndao commented:
I did, it's updated to 0.10.0, unless I missed something.

Wikilicious commented on tests/test_devices/test_gmcse.py:
Though I didn't add linter tests for unittests... it's still good practice to have blank line at the end.

huyndao replied:
I see, I always use black on save and it actually removed the blank line at the end.  However, I can make another commit if you don't mind doing another round of review lol.

huyndao replied:
woops, nvm, looks like you already merged!

Wikilicious reviewed (approved):
Awesome!

Wikilicious commented:
I'm seeing v0.10.0 on pypi... https://pypi.org/project/pygmc/

Wikilicious commented:
> Hi @Wikilicious, thank you for building this nice project and framework.
> 
> I updated it with support for the GMC-SE. It uses the RFC1201 so was fairly straightforward. I've tested it with my GMC-SE and everything seems to work ok. The only exception being that the Gyro gives a tuple of 3 numbers something like (15375, -5885, 769) however I think it is nonsensical and the numbers barely change when the device is moved.
> 
> I've updated the documentations as well. Will create more PR in the future if I see any significant changes needed.
> 
> Huy

For the gyro... The docs don't specify any units so treat it as relative...
Put the GC upside-down and get_gyro

huyndao commented:
> > Hi @Wikilicious, thank you for building this nice project and framework.
> > I updated it with support for the GMC-SE. It uses the RFC1201 so was fairly straightforward. I've tested it with my GMC-SE and everything seems to work ok. The only exception being that the Gyro gives a tuple of 3 numbers something like (15375, -5885, 769) however I think it is nonsensical and the numbers barely change when the device is moved.
> > I've updated the documentations as well. Will create more PR in the future if I see any significant changes needed.
> > Huy
> 
> For the gyro... The docs don't specify any units so treat it as relative... Put the GC upside-down and get_gyro

The change was quite small.  It would be nice if they document what the units are.

- Screen up: (15370, -49, 768)
- Screen down: (15364, -49, 768)
- Most other orientations give the same last 2 numbers

When I have more time, I'll play with it more but it looks like the numbers may change after some event like a reset or history erasure.

Wikilicious commented:
Firstly... Thanks for your contribution!

Second...
On my GMC-500+ 
up-right: (5, -246, 5)
face-down: (8, -10, 230)
face-up: (11, -15, -230)

your screen up/down numbers don't seem to indicate a useful orientation...
Specs I used: https://www.gqelectronicsllc.com/download/GQ-RFC1201.txt

It may be it's just not supported on GMC-SE... it may be bad pygmc coding... it may be bad usb/device...

huyndao commented:
No problems, thank you for creating a convenient package/framework for it.  It will be very helpful for some future projects I have in mind later.

Regarding the GMC-SE: according to this page ![](https://www.gqelectronicsllc.com/support/GMC_Selection_Guide.htm) there is gyro support.  Hopefully I'll have some time this weekend to just connect to it with raw serial and see what is going on.
